src/topology_cacher.py

```diff
@@ -0,0 +1,583 @@
+
+
+resource_info_lookups.json looks like:
+
+```
+{
+  "resource_lists_by_group": {
+    "AGLT2": [
+      {
+        "fqdn": "squid.aglt2.org",
+        "group_name": "AGLT2",
+        "name": "AGLT2-squid",
+        "service_ids": [
+          "138"
+        ],
+        "tags": []
+      },
+      {
+        "fqdn": "sl-um-es3.slateci.io",
+        "group_name": "AGLT2",
+        "name": "AGLT2-squid-2",
+        "service_ids": [
+          "138"
+        ],
+        "tags": []
+      },
+      ...
+    "AMNH": [
+      {
+        "fqdn": "hosted-ce22.opensciencegrid.org",
+        "group_name": "AMNH",
+        "name": "AMNH-ARES",
+        "service_ids": [
+          "1"
+        ],
+        "tags": [
+          "CC*"
+        ]
+      },
+      ...
+    ],
+    ...
+  },
+  "resources_by_fqdn": {
+    ...
+    "249cc.yeg.rac.sh": {
+      "fqdn": "249cc.yeg.rac.sh",
+      "group_name": "CyberaEdmonton",
+      "name": "CYBERA_EDMONTON",
+      "service_ids": [
+        "1"
+      ],
+      "tags": []
+    },
+    "40.119.41.40": {
+      "fqdn": "40.119.41.40",
+      "group_name": "UCSDT2",
+      "name": "UCSDT2-Cloud-3-squid",
+      "service_ids": [
+        "138"
+      ],
+      "tags": []
+    },
+    ...
+  },
+  "resources_by_name": {
+    "AGLT2-squid": {
+      "fqdn": "squid.aglt2.org",
+      "group_name": "AGLT2",
+      "name": "AGLT2-squid",
+      "service_ids": [
+        "138"
+      ],
+      "tags": []
+    },
+    "AGLT2-squid-2": {
+      "fqdn": "sl-um-es3.slateci.io",
+      "group_name": "AGLT2",
+      "name": "AGLT2-squid-2",
+      "service_ids": [
+        "138"
+      ],
+      "tags": []
+    },
+    ...
+  }
+}
+```
+There's some redundancy in the information (e.g. fqdn is included in the
+`resources_by_fqdn` entries) but having a consistent entry format makes
+things easier to read (and was easier to implement).
+
+service_ids are numeric but if we're committed to never changing service
+names then I could make them text instead.
+
+"""
+from argparse import ArgumentParser
+from collections import namedtuple
+import json
+import logging
+import os
+import sys
+import xml.etree.ElementTree as ET
+
+
+from urllib.request import urlopen
+
+from typing import Optional
+
+
+TOPOLOGY = "https://topology.opensciencegrid.org"
+
+
+log = logging.getLogger(__name__)
+
+
+class DataError(Exception):
+    pass
+
+
+class ResourceInfo(namedtuple("ResourceInfo", "group_name name fqdn service_ids tags")):
+    SERVICE_ID_CE = "1"
+    SERVICE_ID_SCHEDD = "109"
+
+    def is_ce(self):
+        return self.SERVICE_ID_CE in self.service_ids
+
+    def is_schedd(self):
+        return self.SERVICE_ID_SCHEDD in self.service_ids
+
+    def to_dict(self) -> dict:
+        return dict(
+            group_name=self.group_name,
+            name=self.name,
+            fqdn=self.fqdn,
+            service_ids=self.service_ids,
+            tags=self.tags,
+        )
```

Comment:
is there a reason why you implement `to_dict` here, rather than using the `_asdict` method that comes with `namedtuple` and as far as i can tell accomplishes the same thing?

Reply:
Did not know about that. Thanks, I was wondering why I had to write that myself...